Support mapping globally ignored permission for a workflow

When a permission is ignored by default it cannot be mapped for a specific workflow.
We should add support for this use case:

``` python


    def test_manage_globally_ignored_permission_for_workflow(self):
        self.load_map_permissions_zcml(
            '<lawgiver:ignore',
            '    permissions="Access contents information" />',

            '<lawgiver:map_permissions',
            '    workflow="foo"'
            '    action_group="view"',
            '    permissions="Access contents information" />',
            )

        registry = self.get_registry()

        self.assertEqual(
            {'group': 'view',
             'ignored': []},

            {'group': registry.get_action_group_for_permission(
                    'Access contents information', workflow_name='foo'),
             'ignored': registry.get_ignored_permissions('foo')})

```

---

```
Failure in test test_manage_ignored_permission_for_workflow (ftw.lawgiver.tests.test_actiongroups.TestActionGroupRegistry)
Traceback (most recent call last):
  File "/Users/jone/projects/python/cache/eggs/unittest2-0.5.1-py2.7.egg/unittest2/case.py", line 340, in run
    testMethod()
  File "/Users/jone/projects/python/cache/eggs/mocker-1.1.1-py2.7.egg/mocker.py", line 149, in test_method_wrapper
    result = test_method()
  File "/Users/jone/projects/packages/ftw.lawgiver/ftw/lawgiver/tests/test_actiongroups.py", line 308, in test_manage_ignored_permission_for_workflow
    'ignored': registry.get_ignored_permissions('foo')})
  File "/Users/jone/projects/python/cache/eggs/unittest2-0.5.1-py2.7.egg/unittest2/case.py", line 521, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/jone/projects/python/cache/eggs/unittest2-0.5.1-py2.7.egg/unittest2/case.py", line 844, in assertDictEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/Users/jone/projects/python/cache/eggs/unittest2-0.5.1-py2.7.egg/unittest2/case.py", line 415, in fail
    raise self.failureException(msg)
AssertionError: {'ignored': [], 'group': 'view'} != {'ignored': set([u'Access contents information']), 'group': None}
- {'group': 'view', 'ignored': []}
+ {'group': None, 'ignored': set([u'Access contents information'])}
```

--- a/docs/HISTORY.txt
+++ b/docs/HISTORY.txt
@@ -5,6 +5,9 @@ Changelog
 1.2 (unreleased)
 ----------------
 
+- Support remapping permission per workflow when it is ignored globally. #22
+  [jone]
+
 - **Support for configuring visible roles in sharing view.**
   The specification now allows to define a list of roles which
   should appear one the sharing view.

--- a/ftw/lawgiver/actiongroups.py
+++ b/ftw/lawgiver/actiongroups.py
@@ -57,6 +57,13 @@ class ActionGroupRegistry(object):
         return permission.get(workflow_name, permission.get(None, None))
 
     def get_ignored_permissions(self, workflow_name=None):
-        permissions = set(self._ignores[None])
-        permissions.update(self._ignores[workflow_name])
+        globally_ignored_permissions = set(self._ignores[None])
+        permissions_ignored_by_workflow = set(self._ignores[workflow_name])
+
+        permissions_managed_by_workflow = set(
+            [permission for permission, workflows in self._permissions.items()
+             if workflow_name in workflows])
+
+        permissions = globally_ignored_permissions - permissions_managed_by_workflow
+        permissions.update(permissions_ignored_by_workflow)
         return permissions

--- a/ftw/lawgiver/tests/test_actiongroups.py
+++ b/ftw/lawgiver/tests/test_actiongroups.py
@@ -322,3 +322,46 @@ class TestActionGroupRegistry(BaseTest):
             'view',
             registry.get_action_group_for_permission(
                 'List folder contents'))
+
+    def test_remap_ignored_permissions_for_a_workflow(self):
+        self.load_map_permissions_zcml(
+            '<lawgiver:ignore',
+            '    permissions="Access contents information" />',
+
+            '<lawgiver:map_permissions',
+            '    action_group="view"',
+            '    workflow="my_workflow"',
+            '    permissions="Access contents information" />',
+            )
+
+        registry = self.get_registry()
+
+        self.assertEqual(
+            {'group': u'view',
+             'ignored': set([]),
+             'view_permissions': set([u'Access contents information'])},
+
+            {'group': registry.get_action_group_for_permission(
+                    'Access contents information', workflow_name='my_workflow'),
+             'ignored': registry.get_ignored_permissions('my_workflow'),
+             'view_permissions': (
+                    registry.get_action_groups_for_workflow(
+                        'my_workflow').get('view'))},
+
+            'Remapping a globally ignored permission to an action group for a'
+            ' specific workflow should make it be re-managed and in the right'
+            ' action group for this workflow, but it does not.')
+
+        self.assertEqual(
+            {'group': None,
+             'ignored': set([u'Access contents information']),
+             'view_permissions': None},
+
+            {'group': registry.get_action_group_for_permission(
+                    'Access contents information', workflow_name='default'),
+             'ignored': registry.get_ignored_permissions('default'),
+             'view_permissions': (
+                    registry.get_action_groups_for_workflow('default').get('view'))},
+
+            'Remapping a globally ignored permission to an action group for a'
+            ' specific workflow should keep it ignored for other workflows')

